src/lib.rs

```diff
@@ -1,271 +1,353 @@
+                <ThinData<Head, SliceItem> as core::fmt::Debug>::fmt(self, f)
             }
         }
-        impl $(<$T>)? core::fmt::Display for $ty $(<$T>)? where $raw: core::fmt::Display, $($($tt)*)? {
-            fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
-                <$raw as core::fmt::Display>::fmt(self, f)
-            }
+
+        unsafe impl<Head, SliceItem> Send for $thin<Head, SliceItem> where
+            $fat<ThinData<Head, SliceItem>>: Send
+        {
         }
-        impl<$($T,)? O: ?Sized> cmp::PartialOrd<O> for $ty $(<$T>)? where $raw: cmp::PartialOrd<O>, $($($tt)*)? {
-            fn partial_cmp(&self, other: &O) -> Option<cmp::Ordering> {
-                <$raw as cmp::PartialOrd<O>>::partial_cmp(self, other)
-            }
+        unsafe impl<Head, SliceItem> Sync for $thin<Head, SliceItem> where
+            $fat<ThinData<Head, SliceItem>>: Sync
+        {
+        }
+
+        impl<Head, SliceItem> cmp::Eq for $thin<Head, SliceItem> where
+            ThinData<Head, SliceItem>: cmp::Eq
+        {
         }
-        impl<$($T,)? O: ?Sized> cmp::PartialEq<O> for $ty $(<$T>)? where $raw: cmp::PartialEq<O>, $($($tt)*)? {
-            fn eq(&self, other: &O) -> bool {
-                <$raw as cmp::PartialEq<O>>::eq(self, other)
+        impl<Head, SliceItem> cmp::PartialEq for $thin<Head, SliceItem>
+        where
+            ThinData<Head, SliceItem>: cmp::PartialEq,
+        {
+            fn eq(&self, other: &Self) -> bool {
+                <ThinData<Head, SliceItem> as cmp::PartialEq>::eq(self, other)
             }
         }
     };
 }
 
-// FUTURE: I think it requires specialization? Get PartialEq/PartialOrd/Eq/Ord for `thin_holder!`s
-macro_rules! total_std_traits {
-    (for $ty:ident $(<$T:ident>)? as $raw:ty $(where $($tt:tt)*)?) => {
-        std_traits!(for $ty $(<$T>)? as $raw $(where $($tt:tt)*)?);
+pub struct ThinBox<Head, SliceItem> {
+    raw: ErasedPtr,
+    marker: PhantomData<Box<ThinData<Head, SliceItem>>>,
+}
+
+thin_holder!(for ThinBox<Head, SliceItem> as Box<ThinData<Head, SliceItem>> with fatten_mut);
+
+impl<Head, SliceItem> ThinBox<Head, SliceItem> {
+    fn layout(len: usize) -> Result<(Layout, [usize; 3]), LayoutErr> {
+        let length_layout = Layout::new::<usize>();
+        let head_layout = Layout::new::<Head>();
+        let slice_layout = layout_array::<SliceItem>(len)?;
+        repr_c_3([length_layout, head_layout, slice_layout])
+    }
+
+    fn alloc(len: usize, layout: Layout) -> ptr::NonNull<ThinData<Head, SliceItem>> {
+        unsafe {
+            let ptr: ErasedPtr = ptr::NonNull::new(alloc(layout))
+                .unwrap_or_else(|| handle_alloc_error(layout))
+                .cast();
+            ptr::write(ThinData::<Head, SliceItem>::length(ptr).as_ptr(), len);
+            ThinData::fatten_mut(ptr.cast())
+        }
+    }
+
+    pub fn new(head: Head, slice: Vec<SliceItem>) -> Self {
```

Comment:
I think servo's signature of `impl Itteraot<Item = SliceItem> + ExactSizeIterator` would be better here (we'll need to check that the iterator's lenght is not a lie, but that should be straighforwrd)

Reply:
`Itteraot`? 😆

There's two issues that mean I think just using `Vec` is fine.

1. `Box<[_]>` doesn't actually implement `IntoIterator`. If you want to into_iter a boxed slice, you need to `Vec::from` it (which is zero cost).
2. By using `Vec` here, we get to use `ptr::copy_nonoverlapping` rather than iterating the source slice. This makes the code not have to worry about panics in the iterator and also guarantees that it's just a single copy rather than piecewise.
3. I'd personally use `TrustedLen` for `new` if it took `ExactSizeIterator`. Perhaps we could offer a secondary constructor that takes an `ExactSizeIterator + ?TrustedLen` if we find it's desired?

Can you think of any use cases in rust-analyzer/rowan that would actually use `ThinArc::new` with a slice that isn't `Vec<_>` or `Box<[_]>`?

All of the places that currently `collect::<Box[_]>` in ra go through `fn with_children(&SyntaxNode, Box<[GreenElement]>)`, and that requires iterating through `new_children` to get the text length as well as passing to `GreenNode::new`.

The "best" signature that `with_children` could take that accepts all of its callers without `collect` would be

```
fn with_children<I>(parent: &SyntaxNode, new_children: I)
where
    I: Iterator<Item=GreenElement> + ExactSizeIterator + TrustedLen,
    I: Clone,
```

and that's only if `fn insert_children` goes from taking `to_insert: &mut dyn Iterator<Item=SyntaxElement>` to `to_insert: &mut dyn CloneableIterator<Item=SyntaxElement>` (where `trait CloneableIterator: Iterator + Clone; impl CloneableIterator for impl Iterator + Clone;`).

TL;DR I don't think it's worth taking something other than `Vec` here.

Reply:
>Can you think of any use cases in rust-analyzer/rowan that would actually use ThinArc::new with a slice that isn't Vec<_> or Box<[_]>?

See https://github.com/rust-analyzer/rowan/pull/36. Basically, green builder already allocates all the nodes in a single vec, and it would be cool if we can drain that vec into a green node without extra allocation. 

Hm, oh, I guess we still have to do one allocation because of an intermediate Box =/

Hm, ok, I think a `Vec` is fine then, though `ExactSizeIterator` does look like a canonical signature here///

Reply:
I've switched it over to take `IntoIterator`. You've won me over and it nicely supports that use case, and is forward-good with directly allocating `Arc`s.

I'm tempted to add a `--cfg abandon-morals` that does it anyway 😛